src/main/java/gregtech/api/recipes/RecipeMap.java

```diff
@@ -228,46 +249,106 @@ public Recipe findRecipe(long voltage, IItemHandlerModifiable inputs, IMultipleT
+        if (recipeSet.isEmpty())
             return null;
         if (minFluidInputs > 0 && GTUtility.amountOfNonNullElements(fluidInputs) < minFluidInputs) {
             return null;
         }
         if (minInputs > 0 && GTUtility.amountOfNonEmptyStacks(inputs) < minInputs) {
             return null;
         }
-        if (maxInputs > 0) {
-            return findByInputs(voltage, inputs, fluidInputs, matchingMode);
-        } else {
-            return findByFluidInputs(voltage, inputs, fluidInputs, matchingMode);
-        }
+        return findByInputsAndFluids(voltage, inputs, fluidInputs, matchingMode, exactVoltage);
     }
 
     @Nullable
-    private Recipe findByFluidInputs(long voltage, List<ItemStack> inputs, List<FluidStack> fluidInputs, MatchingMode matchingMode) {
-        for (FluidStack fluid : fluidInputs) {
-            if (fluid == null) continue;
-            Collection<Recipe> recipes = recipeFluidMap.get(new FluidKey(fluid));
-            if (recipes == null) continue;
-            for (Recipe tmpRecipe : recipes) {
-                if (tmpRecipe.matches(false, inputs, fluidInputs, matchingMode)) {
-                    return voltage >= tmpRecipe.getEUt() ? tmpRecipe : null;
+    private Recipe findByInputsAndFluids(long voltage, List<ItemStack> inputs, List<FluidStack> fluidInputs, MatchingMode matchingMode, boolean exactVoltage) {
+        HashSet<Recipe> iteratedRecipes = new HashSet<>();
+        HashSet<ItemStackKey> searchedItems = new HashSet<>();
+        HashSet<FluidKey> searchedFluids = new HashSet<>();
+        HashMap<Integer, LinkedList<Recipe>> priorityRecipeMap = new HashMap<>();
+        HashMap<Recipe, Integer> promotedTimes = new HashMap<>();
+
+        for (ItemStack stack : inputs) {
+            if (!stack.isEmpty()) {
+                ItemStackKey itemStackKey = KeySharedStack.getRegisteredStack(stack);
+                if (!searchedItems.contains(itemStackKey) && recipeItemMap.containsKey(itemStackKey)) {
+                    searchedItems.add(itemStackKey);
+                    for (Recipe tmpRecipe : recipeItemMap.get(itemStackKey)) {
+                        if (!exactVoltage && voltage < tmpRecipe.getEUt()) {
+                            continue;
+                        } else if (exactVoltage && voltage != tmpRecipe.getEUt()) {
+                            continue;
+                        }
+                        calculateRecipePriority(tmpRecipe, promotedTimes, priorityRecipeMap);
+                    }
                 }
             }
         }
-        return null;
-    }
 
-    @Nullable
-    private Recipe findByInputs(long voltage, List<ItemStack> inputs, List<FluidStack> fluidInputs, MatchingMode matchingMode) {
-        for (Recipe recipe : recipeList) {
-            if (recipe.matches(false, inputs, fluidInputs, matchingMode)) {
-                return voltage >= recipe.getEUt() ? recipe : null;
+        for (FluidStack fluidStack : fluidInputs) {
```

Comment:
Can we short circuit this fluid promotion even more by checking if `matchingMode` is set to `IGNORE_FLUIDS` and skipping the promotion if it is?